### PR TITLE
YARN-10550. Decouple NM runner logic from SLSRunner

### PR DESCRIPTION
**Pull request by szilard-nemeth**
### For code changes:

- [ ] Does the title or this PR starts with the corresponding JIRA issue id (e.g. 'HADOOP-17799. Your PR title ...')?
- [ ] Object storage: have the integration tests been executed and the endpoint declared according to the connector-specific documentation?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the `LICENSE`, `LICENSE-binary`, `NOTICE-binary` files?

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 52s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
|||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  37m 26s |  |  trunk passed  |
| +1 :green_heart: |  compile  |   0m 28s |  |  trunk passed with JDK Ubuntu-11.0.14+9-Ubuntu-0ubuntu2.20.04  |
| +1 :green_heart: |  compile  |   0m 23s |  |  trunk passed with JDK Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07  |
| +1 :green_heart: |  checkstyle  |   0m 21s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   0m 28s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   0m 26s |  |  trunk passed with JDK Ubuntu-11.0.14+9-Ubuntu-0ubuntu2.20.04  |
| +1 :green_heart: |  javadoc  |   0m 24s |  |  trunk passed with JDK Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07  |
| -1 :x: |  spotbugs  |   0m 50s | [/branch-spotbugs-hadoop-tools_hadoop-sls-warnings.html](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4116/1/artifact/out/branch-spotbugs-hadoop-tools_hadoop-sls-warnings.html) |  hadoop-tools/hadoop-sls in trunk has 2 extant spotbugs warnings.  |
| +1 :green_heart: |  shadedclient  |  23m 43s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 21s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 21s |  |  the patch passed with JDK Ubuntu-11.0.14+9-Ubuntu-0ubuntu2.20.04  |
| +1 :green_heart: |  javac  |   0m 21s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 18s |  |  the patch passed with JDK Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07  |
| +1 :green_heart: |  javac  |   0m 18s |  |  the patch passed  |
| -1 :x: |  blanks  |   0m  0s | [/blanks-eol.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4116/1/artifact/out/blanks-eol.txt) |  The patch has 5 line(s) that end in blanks. Use git apply --whitespace=fix <<patch_file>>. Refer https://git-scm.com/docs/git-apply  |
| -0 :warning: |  checkstyle  |   0m 12s | [/results-checkstyle-hadoop-tools_hadoop-sls.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4116/1/artifact/out/results-checkstyle-hadoop-tools_hadoop-sls.txt) |  hadoop-tools/hadoop-sls: The patch generated 21 new + 54 unchanged - 6 fixed = 75 total (was 60)  |
| +1 :green_heart: |  mvnsite  |   0m 21s |  |  the patch passed  |
| -1 :x: |  javadoc  |   0m 19s | [/results-javadoc-javadoc-hadoop-tools_hadoop-sls-jdkUbuntu-11.0.14+9-Ubuntu-0ubuntu2.20.04.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4116/1/artifact/out/results-javadoc-javadoc-hadoop-tools_hadoop-sls-jdkUbuntu-11.0.14+9-Ubuntu-0ubuntu2.20.04.txt) |  hadoop-tools_hadoop-sls-jdkUbuntu-11.0.14+9-Ubuntu-0ubuntu2.20.04 with JDK Ubuntu-11.0.14+9-Ubuntu-0ubuntu2.20.04 generated 1 new + 25 unchanged - 0 fixed = 26 total (was 25)  |
| -1 :x: |  javadoc  |   0m 17s | [/results-javadoc-javadoc-hadoop-tools_hadoop-sls-jdkPrivateBuild-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4116/1/artifact/out/results-javadoc-javadoc-hadoop-tools_hadoop-sls-jdkPrivateBuild-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07.txt) |  hadoop-tools_hadoop-sls-jdkPrivateBuild-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07 with JDK Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07 generated 1 new + 25 unchanged - 0 fixed = 26 total (was 25)  |
| -1 :x: |  spotbugs  |   0m 55s | [/new-spotbugs-hadoop-tools_hadoop-sls.html](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4116/1/artifact/out/new-spotbugs-hadoop-tools_hadoop-sls.html) |  hadoop-tools/hadoop-sls generated 1 new + 2 unchanged - 0 fixed = 3 total (was 2)  |
| +1 :green_heart: |  shadedclient  |  23m 25s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |  12m 16s |  |  hadoop-sls in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 32s |  |  The patch does not generate ASF License warnings.  |
|  |   | 105m 45s |  |  |


| Reason | Tests |
|-------:|:------|
| SpotBugs | module:hadoop-tools/hadoop-sls |
|  |  org.apache.hadoop.yarn.sls.NMRunner.setInputTraces(String[]) may expose internal representation by storing an externally mutable object into NMRunner.inputTraces  At NMRunner.java:by storing an externally mutable object into NMRunner.inputTraces  At NMRunner.java:[line 208] |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4116/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/4116 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell |
| uname | Linux d436d1eb068f 4.15.0-163-generic #171-Ubuntu SMP Fri Nov 5 11:55:11 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / 5780f12441cd71a78166d62ec7cc98d50f512120 |
| Default Java | Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.14+9-Ubuntu-0ubuntu2.20.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4116/1/testReport/ |
| Max. process+thread count | 561 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-sls U: hadoop-tools/hadoop-sls |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4116/1/console |
| versions | git=2.25.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.14.0-SNAPSHOT https://yetus.apache.org |


This message was automatically generated.

File: hadoop-tools/hadoop-sls/src/main/java/org/apache/hadoop/yarn/sls/scheduler/SchedulerMetrics.java [hunk 1]
```diff
@@ -362,6 +362,10 @@ public Integer getValue() {
     );
   }
 
+  private boolean isMetricsAvailable() {
+    return scheduler.getRootQueueMetrics() == null;
+  }
+
```
Comment (9uapaw) [inline hadoop-tools/hadoop-sls/src/main/java/org/apache/hadoop/yarn/sls/scheduler/SchedulerMetrics.java:368]: I do not think it is a good idea to introduce a non-SLS change in this patch. I would refrain from changing any production code for the sake of SLS.

File: hadoop-tools/hadoop-sls/src/main/java/org/apache/hadoop/yarn/sls/AMRunner.java [hunk 2]
```diff
@@ -148,16 +147,15 @@ private void startAMFromSLSTrace(String inputTrace) throws IOException {
   private void startAMFromSynthGenerator() throws YarnException, IOException {
     Configuration localConf = new Configuration();
     localConf.set("fs.defaultFS", "file:///");
-    // if we use the nodeFile this could have been not initialized yet.
-    if (stjp == null) {
-      stjp = new SynthTraceJobProducer(conf, new Path(inputTraces[0]));
-      slsRunner.setStjp(stjp);
+    //if we use the nodeFile this could have been not initialized yet.
+    if (slsRunner.getStjp() == null) {
+      slsRunner.setStjp(new SynthTraceJobProducer(conf, new Path(inputTraces[0])));
```
Comment (9uapaw) [inline hadoop-tools/hadoop-sls/src/main/java/org/apache/hadoop/yarn/sls/AMRunner.java:152]: This looks like SLSRunner#getSynthJobTraceProducer.

**Review by 9uapaw — COMMENTED** [review]
Thanks for the change @szilard-nemeth, generally LGTM, but I had some minor comments.

File: hadoop-tools/hadoop-sls/src/main/java/org/apache/hadoop/yarn/sls/scheduler/SchedulerMetrics.java (hunk 1 above)
Comment (9uapaw) [reply hadoop-tools/hadoop-sls/src/main/java/org/apache/hadoop/yarn/sls/scheduler/SchedulerMetrics.java:368]: Nevermind, its SLS.

**Review by brumi1024 — APPROVED** [review]
@szilard-nemeth thanks for the patch. Since the spotbugs issue will be handled in YARN-11102 and the rest of the issues are just moved and a generic cleanup is in progress (YARN-11103) so +1 from my side.

**Comment by brumi1024** [comment]
Thanks @szilard-nemeth, committed to trunk.

**Pull request closed without merging**